### DISCUSSION TITLE
fl_imgLdr progress

### DISCUSSION THREAD

**Discussion by rageworx** [fire-eggs/YAIV#85] [General]
This is an utility feature to sense image automatically and loads on Fl_RGB_Image or Fl_Animated_Image ( new feature own render pthread ).
I will share here what progress is going on.

## Planned feature
- Multiplatform ( except MSVC, it has too many compatibility form MSVC features )
- Loads any supported image from memory
- Loads any supported image from file system in UTF-8 ( maybe Windows need to use fl_utf8xx features to convert encodings )
- Automatically rotated or color calibration from EXiF meta informations
- New feature for Fl_Animated_Image, it has own pthread loop to update own render source ( Fl_RGB_Image ) and calling itself to update image like fl_callback.

**Comment by rageworx** [comment]
## Progress until Aug.02.2021
- I was too busy to making new works for my job - new hardware locked licenses ( Windows, Mac and Linux ) and developing new Embedde linux peripheral USB3.1 Gen1 super speed protocol ... 
- Used libexif to recognize what meta information contained for PNG and JPG ( or other image format existed to contain EXiF meta tags ? )
- For a while, libexif updated and I couldn't catch what changes by security vulnerable issue.
- Yeseterday I had fixed this security issue , now it sense exif meta information again.
- Image can load faster ( by using libturbojpeg, not libjpeg-turbo, it's different one )
- Loaded image automatically rotated or flipped by EXiF meta.

**Reply by fire-eggs** [reply]
Thanks for the update! Work can be so distracting ;-}

I did not know about libturbojpeg vs libjpeg-turbo - cool!

**Comment by rageworx** [comment]
## Developing environment
- MSYS2 + MinGW-W64
- MacOS 10.14 and 11.5 + iTerm2 + brew + xcode console, I'm using MacBook Pro 2015 ratina 13", MacBookPro 2017 13" and MacBook Air 13"(M1) for now.
- Linux ( debian, x86.64 and aarch64 ), LinuxMint18.x and Linaro for my RK3399 embedded board.
- Sorry MSVC is not an option for now ...

**Comment by rageworx** [comment]
## Question
- Which method is better for YAIV ?
- Fl_Animated_Image has virutal class implementation to update image?
- Or calling fl_callback to redraw/update latest changed image ?
- Both need pthread for now ( or better than to use std:thread ? it's slower than pthread but easy to port to most of C++11 compilers )

**Reply by fire-eggs** [reply]
These are good questions. Sorry if the following is disorganized ... you touched on something I've been struggling with!

I'm not entirely happy with the way `Fl_Anim_GIF_Image` works.

`Fl_Anim_GIF_Image` takes a `Fl_Widget` "canvas" and uses `Fl_Widget::image()` to display the image. The class has its own timer and updates the "canvas" when necessary. `XBox` has to detach the canvas when it's time to stop playing the animated image. This also means that `XBox` can't use any alternative drawing mechanism for animated images.

YAIV (specifically `XBox`) wants to do the drawing so that scale / dither / checker / transforms can be applied. Currently XBox has to treat animated images "special": look for all the checks against `_anim` in XBox.cpp. At the moment, dither and transforms aren't supported for animated images because of this. If possible I'd like to have the current-image / current-frame-image be handled the same.

[Note this could also mean that `XBox` is just badly designed. Ideas welcome... ]

I think I'd like to see `Fl_Animated_Image` as a multi-image container, with access to individual frames. It might make more sense to have `Fl_Animated_Image` provide the timer with a callback, so that `XBox` doesn't have to have smarts about things like "initial delay", "frame delay", "loop count", and so on.
 
[Not relevant to the questions, but I'd like to have a "playback control" which lets the user start / stop playback, step through frames, start over, etc. Should be feasible by asking `Fl_Animated_Image` to stop the timer playback.]

I wonder if two classes are necessary: 1) `Fl_Multi_Image` and 2) `Fl_Animated_Image`. The first class is a container of multiple images, useful for animation _and_ multi-page formats like TIF. The second derives from the first and adds the timing data, etc. This is very optional at this point!
I'm not sure where the "cached" images should be stored [as playback happens, need to cache the scaled / dithered / transformed frames rather than repeat the processing every time ...]

[re: `pthread`, I've used `_beginthread` successfully on Windows to replace `pthread_create`. Not sure about the implications for other `pthread` functions... or the timer ... ]

**Reply by rageworx** [reply]
Ok, I mean, acutally, my Fl_Animated_Image is not same with your GIF animated widget.
It's very simple designed new image widget contains thread - but it's little bit dangerous decision by each one thread for any animation image - is drains system thread resources, so I'm still making it how works more efficiently for less system resources.
YAIV maybe just doing this for Fl_Animated_Image
- Fl_RGB_Image* GetElapsedImage( timeval* tv );
- Fl_RGB_Image* GetElapsedImage( timeval* fromtv, timeval* nowtv );
It means, all animated images has own time based frames, and Fl_Animated_Image returns duplicated new Fl_RGB_Image* with GetEleapsedImage() with timeval - and using ms or sec. by calculating how many times ran, and exactly timed image should got by each interval by YAIV - yes, there's nothing like canvas.
And frame timing may adjusted by this method,
- SetTimeRate( float tr );
- ResetTimeRate();
Timerate is a ratio by multiply, less 1.0 meaning ( original time rate * 0.8 ) slower, and higher than 1.0 meaning faster.
It is not compatible with original Fl_Animated_Image.
I will serve codes and examples as I can as soon.
Is this good for you ?

**Reply by fire-eggs** [reply]
Looking forward to it! I'm currently in a slump, this might get me motivated again...

**Reply by rageworx** [reply]
> Looking forward to it! I'm currently in a slump, this might get me motivated again...

Oh, slump, it's an enemy's name of all engineers.
I'm sorry to hear that ...

Actually I'm also in slump too, maybe exhausted temporary.
Too many things to do, working from home, lazy fellow workers … too many things.
I believe you already experienced some swindlers in anywhere, who make lies plausible, they make situation more critical, even with Incompetent director, today's feels like HELL.
I will back to here asap, solve some critical issues ...

**Reply by fire-eggs** [reply]
>Actually I'm also in slump too, maybe exhausted temporary.
>Too many things to do, working from home, lazy fellow workers … too many things.

Take care of yourself - burnout is a real thing! It was one of the factors that led me to retire early.

Any one of those issues can be enough to cause a slump - I'm sorry to hear you're experiencing so many!!
